src/model/Graph.cpp

```diff
@@ -406,37 +428,184 @@ pair<vector<Node*>,double> Graph:: tspTriangularApproximation() {
 
     // Add the start node to complete the tour
     tspTour.push_back(startNode);
-    double weight=0;
-    for(int i=0; i<tspTour.size()-1;i++){
-        weight+= calculateDistance(tspTour[i], tspTour[i+1]);
+    double weight = 0;
+    for (int i = 0; i < tspTour.size() - 1; i++)
+    {
+        weight += calculateDistance(tspTour[i], tspTour[i + 1]);
+        tspTour[i]->tspIndex=i;
     }
+    pair<vector<Node*>,double> tsp={tspTour, weight};
 
-    return {tspTour, weight};
+    return tsp;
 }
+void Graph::greddyImprovement(bool* run,double *solution,pair<vector<Node*>,double> &tsp){
+    vector<Edge> edgesVector;
+    for(auto edges_ : edges){
+        for(auto edge: edges_){
+            if(edge!= nullptr)edgesVector.push_back(*edge);
+        }
+    }
+    sort(edgesVector.begin(),edgesVector.end(),[](Edge a, Edge b){
+        return a.getCapacity() > b.getCapacity();
+    });
+    int i=edgesVector.size();
+    while (!edgesVector.empty() and *run){
```

Comment:
you're doing all of these stuff, that takes time and memory, even if run is false. Couldn't you check it outside?